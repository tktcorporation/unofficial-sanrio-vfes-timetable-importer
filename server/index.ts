// server/index.ts
import { Hono } from "hono";
import {
	addToCalendar,
	generateICS,
	generateCancelICS,
	getAuthUrl,
	getEvents,
	handleAuthCallback,
} from "./controller";

const app = new Hono<{
	Bindings: {
		MY_VAR: string;
		GOOGLE_CLIENT_ID: string;
		GOOGLE_CLIENT_SECRET: string;
	};
	Variables: {
		MY_VAR_IN_VARIABLES: string;
	};
}>();

app.use(async (c, next) => {
	c.set("MY_VAR_IN_VARIABLES", "My variable set in c.set");
	await next();
	c.header("X-Powered-By", "React Router and Hono");
});

const routes = app
	.get("/api", (c) => {
		return c.json({
			message: "Hello",
			var: c.env.MY_VAR,
		});
	})
	.get("/events", getEvents)
	.get("/auth/url", getAuthUrl)
	.post("/auth/callback", handleAuthCallback)
	.post("/calendar/add", addToCalendar)
	.post("/calendar/ics", generateICS)
	.post("/calendar/cancel-ics", generateCancelICS);

export type AppType = typeof routes;

export default app;
