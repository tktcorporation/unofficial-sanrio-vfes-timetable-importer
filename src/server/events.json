[
  {
    "platform": ["PC"],
    "title": "AMOKA",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_84f13ef3-98c1-4307-baba-ac3abc48dd22_small.webp",
    "schedules": [
      { "date": "2/9", "time": "19:30" },
      { "date": "3/8", "time": "13:30" },
      { "date": "3/16", "time": "7:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "夕月ティア",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_f0d77f9d-59ac-4428-ba8e-4b914cbc09b5_small.webp",
    "schedules": [
      { "date": "2/9", "time": "20:00" },
      { "date": "3/8", "time": "14:00" },
      { "date": "3/16", "time": "8:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "YSS",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_202f6e41-c7c4-49bb-9870-318cdbe0bde0_small.webp",
    "schedules": [
      { "date": "2/9", "time": "20:30" },
      { "date": "3/8", "time": "14:30" },
      { "date": "3/16", "time": "8:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "フラガリアメモリーズ",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_7b1758a0-1867-4877-902e-94e141f5167f_small.webp",
    "schedules": [
      { "date": "2/15", "time": "18:30" },
      { "date": "3/8", "time": "18:00" },
      { "date": "3/16", "time": "12:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "LuminaPalette featuring ほわん",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_27eb71d8-d2fd-417e-b5f8-0eb83ee92f59_small.webp",
    "schedules": [
      { "date": "2/15", "time": "19:00" },
      { "date": "3/8", "time": "15:00" },
      { "date": "3/16", "time": "9:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "約束 with デルミン",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_fb19efd7-e799-4e4e-b094-2c5598a2c9a7_small.webp",
    "schedules": [
      { "date": "2/15", "time": "19:30" },
      { "date": "3/8", "time": "15:30" },
      { "date": "3/16", "time": "9:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "orange & MIRA_sk featuring サンリオキャラクターズ",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_64669b31-15f1-41f2-9288-146616e07aa3_small.webp",
    "schedules": [
      { "date": "2/15", "time": "20:00" },
      { "date": "3/8", "time": "16:00" },
      { "date": "3/16", "time": "10:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "TOKYO WAIYOZ",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_e2f895b1-e5d1-433a-bedc-3d11a26a732d_small.webp",
    "schedules": [
      { "date": "2/16", "time": "19:00" },
      { "date": "3/9", "time": "15:00" },
      { "date": "3/17", "time": "9:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "exxp",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_ace8d763-a48d-46af-a5cb-14fde4c4f732_small.webp",
    "schedules": [
      { "date": "2/16", "time": "19:30" },
      { "date": "3/9", "time": "15:30" },
      { "date": "3/17", "time": "9:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "因幡はねる",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_ef64651b-6572-4126-8eeb-f990eb7ed070_small.webp",
    "schedules": [
      { "date": "2/16", "time": "20:00" },
      { "date": "3/9", "time": "19:00" },
      { "date": "3/17", "time": "13:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "P丸様。",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_9bb8648e-ce4c-4ec4-8457-e6a9e932021a_small.webp",
    "schedules": [
      { "date": "2/16", "time": "20:30" },
      { "date": "3/9", "time": "19:30" },
      { "date": "3/17", "time": "13:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "ironmouse",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_b10438f8-57c9-4d98-8b8f-168c6e33ace4_small.webp",
    "schedules": [
      { "date": "2/23", "time": ["10:30", "18:30"] },
      { "date": "3/9", "time": "13:30" },
      { "date": "3/17", "time": "9:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "因幡はねる",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_ef64651b-6572-4126-8eeb-f990eb7ed070_small.webp",
    "schedules": [
      { "date": "2/16", "time": "20:00" },
      { "date": "3/9", "time": "19:00" },
      { "date": "3/17", "time": "13:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "P丸様。",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_9bb8648e-ce4c-4ec4-8457-e6a9e932021a_small.webp",
    "schedules": [
      { "date": "2/16", "time": "20:30" },
      { "date": "3/9", "time": "19:30" },
      { "date": "3/17", "time": "13:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "ironmouse",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_b10438f8-57c9-4d98-8b8f-168c6e33ace4_small.webp",
    "schedules": [
      { "date": "2/16", "time": "19:00" },
      { "date": "3/9", "time": "19:30" },
      { "date": "3/17", "time": "13:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "TOKYO WAIYOZ",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_e2f895b1-e5d1-433a-bedc-3d11a26a732d_small.webp",
    "schedules": [
      { "date": "2/16", "time": "19:00" },
      { "date": "3/9", "time": "15:00" },
      { "date": "3/17", "time": "9:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "ironmouse",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_b10438f8-57c9-4d98-8b8f-168c6e33ace4_small.webp",
    "schedules": [
      { "date": "2/23", "time": ["10:30", "18:30"] },
      { "date": "3/9", "time": "13:30" },
      { "date": "3/17", "time": "9:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "Virtual Reverb: The Creative Echo of hide",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-904x508_v-fs_webp_b1675073-b540-45de-8228-b0efab992782_small.webp",
    "schedules": [
      { "date": "3/1", "time": "19:30" },
      { "date": "3/8", "time": "19:00" },
      { "date": "3/16", "time": "10:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "樋口楓",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_5fb0c7d1-65f4-4d76-b59a-704cf52c0762_small.webp",
    "schedules": [
      { "date": "3/1", "time": "20:00" },
      { "date": "3/8", "time": "20:00" },
      { "date": "3/16", "time": "14:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "月ノ美兎",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_e5e8a1bc-ba3a-45cf-8dd1-ed570426ff47_small.webp",
    "schedules": [
      { "date": "3/1", "time": "20:30" },
      { "date": "3/8", "time": "20:30" },
      { "date": "3/16", "time": "14:30" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "Mashumairesh!!(SHOW BY ROCK!!)",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_e845344a-b9df-427d-bdbd-22908f8fcea8_small.webp",
    "schedules": [
      { "date": "3/2", "time": "20:00" },
      { "date": "3/9", "time": "20:00" },
      { "date": "3/17", "time": "14:00" }
    ]
  },
  {
    "platform": ["Android", "PC"],
    "title": "QuestMaker featuring サンリオキャラクターズ",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_8b8943e2-1c07-495d-a553-5a871d27b164_small.webp",
    "schedules": [
      { "date": "3/2", "time": "18:30" },
      { "date": "3/9", "time": "16:00" },
      { "date": "3/17", "time": "10:00" }
    ]
  },
  {
    "platform": ["PC"],
    "title": "キヌ",
    "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_b8ecfeaf-9267-4520-b9e5-b68df428da46_small.webp",
    "schedules": [
      { "date": "3/2", "time": "19:30" },
      { "date": "3/9", "time": "17:00" },
      { "date": "3/17", "time": "11:00" }
    ]
  }
]
