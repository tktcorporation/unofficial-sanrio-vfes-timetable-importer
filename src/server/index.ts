import { readFileSync } from 'fs';
import { join } from 'path';
import { serve } from '@hono/node-server';
import { google } from 'googleapis';
import { Hono } from 'hono';
import { cors } from 'hono/cors';

const app = new Hono();
app.use(cors());

const oauth2Client = new google.auth.OAuth2(
  process.env.GOOGLE_CLIENT_ID,
  process.env.GOOGLE_CLIENT_SECRET,
  'http://localhost:5173/callback'
);

const calendar = google.calendar({ version: 'v3', auth: oauth2Client });

app.get('/events', (c) => {
  const eventsPath = join(process.cwd(), 'src', 'server', 'events.json');
  const events = JSON.parse(readFileSync(eventsPath, 'utf-8'));
  return c.json(events);
});

app.get('/auth/url', (c) => {
  const url = oauth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: ['https://www.googleapis.com/auth/calendar.events'],
  });
  return c.json({ url });
});

app.post('/auth/callback', async (c) => {
  const { code } = await c.req.json();
  const { tokens } = await oauth2Client.getToken(code);
  oauth2Client.setCredentials(tokens);
  return c.json({ success: true });
});

app.post('/calendar/add', async (c) => {
  const { events } = await c.req.json();

  try {
    const results = await Promise.all(
      events.map(async (event: any) => {
        const [year, month, day] = event.date.split('/');
        const [hour, minute] = event.time.split(':');
        const startTime = new Date(
          2024,
          Number.parseInt(month) - 1,
          Number.parseInt(day),
          Number.parseInt(hour),
          Number.parseInt(minute)
        );
        const endTime = new Date(startTime.getTime() + 60 * 60 * 1000); // 1 hour duration

        return calendar.events.insert({
          calendarId: 'primary',
          requestBody: {
            summary: event.title,
            description: `Platform: ${event.platform.join(', ')}`,
            start: {
              dateTime: startTime.toISOString(),
              timeZone: 'Asia/Tokyo',
            },
            end: {
              dateTime: endTime.toISOString(),
              timeZone: 'Asia/Tokyo',
            },
          },
        });
      })
    );
    return c.json({ success: true, results });
  } catch (error) {
    return c.json({ success: false, error: error.message }, 500);
  }
});

const port = process.env.PORT || 3000;
console.log(`Server is running on port ${port}`);

serve({
  fetch: app.fetch,
  port: Number(port),
});
