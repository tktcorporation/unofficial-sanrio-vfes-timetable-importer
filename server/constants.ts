import { z } from "zod";
import { events as _events } from "./events.json";

const eventSchema = z.array(
	z.object({
		uid: z.string().uuid(),
		floor: z.enum(["B4F", "1F/2F", "4F", "B3F", "その他"]),
		ticketLink: z.string().optional(),
		path: z.string().optional(),
		locationName: z.string().optional(),
		description: z.string().optional(),
		platform: z.array(z.enum(["PC", "Android"])),
		title: z.string().min(1),
		image: z.string().optional(),
		timeSlotMinutes: z.number(),
		schedules: z.array(
			z.object({
				year: z.string().regex(/^\d{4}$/),
				date: z.object({
					// 一文字 or 二文字
					month: z.string().regex(/^\d{1,2}$/),
					day: z.string().regex(/^\d{1,2}$/),
				}),
				time: z.object({
					hour: z.string().regex(/^\d{1,2}$/),
					minute: z.string().regex(/^\d{1,2}$/),
				}),
			}),
		),
	}),
);
export const EVENTS: Event[] = eventSchema.parse(_events);
export type Event = z.infer<typeof eventSchema>[number];
