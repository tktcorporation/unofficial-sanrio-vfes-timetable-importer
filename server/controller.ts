import type { Context } from "hono";
import { z } from "zod";
import { events } from "./events.json";

const eventSchema = z.array(
	z.object({
		platform: z.array(z.enum(["PC", "Android"])),
		title: z.string(),
		image: z.string(),
		schedules: z.array(
			z.object({
				date: z.object({
					month: z.string(),
					day: z.string(),
				}),
				time: z.object({
					hour: z.string(),
					minute: z.string(),
				}),
			}),
		),
	}),
);

const calendarEventSchema = z.array(
	z.object({
		title: z.string(),
		date: z.string(),
		time: z.string(),
		platform: z.array(z.enum(["PC", "Android"])),
	}),
);

export const getEvents = async (c: Context) => {
	const validatedEvents = eventSchema.parse(events);
	return c.json(validatedEvents);
};

export const getAuthUrl = async (c: Context) => {
	// TODO: 実際のOAuth設定に応じて変更する必要があります
	const clientId = c.env.GOOGLE_CLIENT_ID;
	const origin = new URL(c.req.url).origin;
	const redirectUri = `${origin}/auth/callback`;
	const scope = "https://www.googleapis.com/auth/calendar";

	const authUrl = `https://accounts.google.com/o/oauth2/v2/auth?client_id=${clientId}&redirect_uri=${encodeURIComponent(redirectUri)}&scope=${encodeURIComponent(scope)}&response_type=code`;

	return c.json({ url: authUrl });
};

export const handleAuthCallback = async (c: Context) => {
	try {
		const { code } = await c.req.json();
		if (!code) {
			return c.json(
				{ success: false, error: "認証コードが見つかりません" },
				400,
			);
		}

		// TODO: 実際のOAuth設定に応じて変更する必要があります
		const clientId = c.env.GOOGLE_CLIENT_ID;
		const clientSecret = c.env.GOOGLE_CLIENT_SECRET;
		const origin = new URL(c.req.url).origin;
		const redirectUri = `${origin}/auth/callback`;

		const tokenResponse = await fetch("https://oauth2.googleapis.com/token", {
			method: "POST",
			headers: {
				"Content-Type": "application/x-www-form-urlencoded",
			},
			body: new URLSearchParams({
				code,
				client_id: clientId,
				client_secret: clientSecret,
				redirect_uri: redirectUri,
				grant_type: "authorization_code",
			}),
		});

		if (!tokenResponse.ok) {
			return c.json(
				{ success: false, error: "トークンの取得に失敗しました" },
				400,
			);
		}

		const tokenData = await tokenResponse.json();
		// TODO: トークンを保存する処理を追加

		return c.json({ success: true });
	} catch (error) {
		console.error("Auth callback error:", error);
		return c.json({ success: false, error: "認証処理に失敗しました" }, 500);
	}
};

export const addToCalendar = async (c: Context) => {
	try {
		const { events } = await c.req.json();
		const validatedEvents = calendarEventSchema.parse(events);

		// TODO: 実際のGoogle Calendar APIを使用してイベントを追加
		// この例では成功レスポンスを返すだけです
		return c.json({ success: true });
	} catch (error) {
		if (error instanceof z.ZodError) {
			return c.json({ success: false, error: "不正なイベントデータです" }, 400);
		}
		console.error("Add to calendar error:", error);
		return c.json(
			{ success: false, error: "カレンダーへの追加に失敗しました" },
			500,
		);
	}
};

export const generateICS = async (c: Context) => {
	try {
		const { events } = await c.req.json();
		const validatedEvents = calendarEventSchema.parse(events);

		// ICSファイルの生成
		const icsContent = [
			"BEGIN:VCALENDAR",
			"VERSION:2.0",
			"PRODID:-//sanrio-vfes-timetable-importer//JP",
			...validatedEvents.map((event) => {
				const [month, day] = event.date.split("/");
				const [hour, minute] = event.time.split(":");
				const dateStr = `2024${month.padStart(2, "0")}${day.padStart(2, "0")}T${hour.padStart(2, "0")}${minute.padStart(2, "0")}00`;

				return [
					"BEGIN:VEVENT",
					`SUMMARY:${event.title}`,
					`DTSTART:${dateStr}`,
					`DTEND:${dateStr}`,
					`DESCRIPTION:プラットフォーム: ${event.platform.join(", ")}`,
					"END:VEVENT",
				].join("\n");
			}),
			"END:VCALENDAR",
		].join("\n");

		return new Response(icsContent, {
			headers: {
				"Content-Type": "text/calendar",
				"Content-Disposition": "attachment; filename=events.ics",
			},
		});
	} catch (error) {
		if (error instanceof z.ZodError) {
			return c.json({ success: false, error: "不正なイベントデータです" }, 400);
		}
		console.error("Generate ICS error:", error);
		return c.json(
			{ success: false, error: "ICSファイルの生成に失敗しました" },
			500,
		);
	}
};
