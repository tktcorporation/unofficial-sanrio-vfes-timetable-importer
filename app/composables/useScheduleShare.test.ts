import { EVENTS } from "server/constants";
import { describe, expect, it } from "vitest";
import {
	compressSchedules,
	dateToJulian,
	decodeBinaryData,
	decompressSchedules,
	encodeBinaryData,
	julianToDate,
} from "./useScheduleShare";

function assertDefined<T>(value: T | undefined): asserts value is T {
	expect(value).toBeDefined();
}

describe("useScheduleShare", () => {
	describe("日付変換", () => {
		it("dateToJulian と julianToDate が相互に変換できること", () => {
			const testDate = {
				year: 2024,
				month: 3,
				day: 15,
			};

			const julian = dateToJulian(testDate);
			const convertedDate = julianToDate(julian);

			expect(convertedDate).toEqual(testDate);
		});

		it("異なる日付で正しく変換できること", () => {
			const testCases = [
				{ year: 2024, month: 1, day: 1 },
				{ year: 2024, month: 12, day: 31 },
				{ year: 2025, month: 6, day: 15 },
			];

			for (const date of testCases) {
				const julian = dateToJulian(date);
				const converted = julianToDate(julian);
				expect(converted).toEqual(date);
			}
		});
	});

	describe("バイナリデータ変換", () => {
		it("encodeBinaryData と decodeBinaryData が相互に変換できること", () => {
			const testData = [
				[1000, 74, 120],
				[2000, 75, 240],
			];

			const encoded = encodeBinaryData(testData);
			const decoded = decodeBinaryData(encoded);

			expect(decoded).toEqual(testData);
		});
	});

	describe("スケジュール圧縮", () => {
		const mockSchedules = [
			{
				uid: "abc123",
				schedule: {
					date: { year: 2024, month: 3, day: 15 },
					time: { hour: 13, minute: 30 },
				},
			},
		];

		it("スケジュールを圧縮できること", () => {
			const compressed = compressSchedules({ schedules: mockSchedules });
			expect(typeof compressed).toBe("string");
			expect(compressed.length).toBeGreaterThan(0);
		});

		it("圧縮したスケジュールを展開できること", () => {
			const compressed = compressSchedules({ schedules: mockSchedules });
			const mockEvents = [{ uid: "abc123" }];

			const decompressed = decompressSchedules({
				compressed,
				events: mockEvents,
			});

			expect(decompressed[0]).toEqual({
				uid: "abc123",
				schedule: {
					date: { year: 2024, month: 3, day: 15 },
					time: { hour: 13, minute: 30 },
				},
			});
		});

		it("無効な圧縮データでエラーを投げること", () => {
			const mockEvents = [{ uid: "abc123" }];
			expect(() => {
				decompressSchedules({ compressed: "invalid-data", events: mockEvents });
			}).toThrow("無効な圧縮データです");
		});

		it("存在しないイベントIDでエラーを投げること", () => {
			const compressed = compressSchedules({ schedules: mockSchedules });
			const mockEvents = [{ uid: "xyz789" }];

			expect(() => {
				decompressSchedules({ compressed, events: mockEvents });
			}).toThrow("該当するイベントが見つかりません");
		});
	});

	describe("実際のデータでテスト", () => {
		it("QM2のデータが正しく圧縮できること", () => {
			const events = EVENTS;
			const QM2 = events.find(
				(e) => e.uid === "5a6fb727-cd94-5647-9329-b0129edfd4e5",
			);
			assertDefined(QM2);
			expect(QM2).toMatchSnapshot();

			const schedules = QM2.schedules.map((s) => ({
				uid: QM2.uid,
				schedule: {
					date: {
						year: Number(s.year),
						month: Number(s.date.month),
						day: Number(s.date.day),
					},
					time: {
						hour: Number(s.time.hour),
						minute: Number(s.time.minute),
					},
				},
			}));

			expect(schedules).toMatchSnapshot();

			const compressed = compressSchedules({ schedules });
			expect(compressed).toMatchSnapshot();

			const decompressed = decompressSchedules({ compressed, events });
			expect(decompressed).toMatchSnapshot();

			expect(decompressed.length).toBe(schedules.length);
			expect(decompressed).toEqual(schedules);
		});
	});
});
