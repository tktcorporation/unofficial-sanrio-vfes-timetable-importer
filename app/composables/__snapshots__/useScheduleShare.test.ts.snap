// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useScheduleShare > 実際のデータでテスト > QM2のデータが正しく圧縮できること 1`] = `
{
  "floor": "B4F",
  "image": "https://storage.googleapis.com/studio-cms-assets/projects/G3qbEkMgOJ/s-1280x720_v-fms_webp_8b8943e2-1c07-495d-a553-5a871d27b164_small.webp",
  "platform": [
    "Android",
    "PC",
  ],
  "schedules": [
    {
      "date": {
        "day": "2",
        "month": "3",
      },
      "time": {
        "hour": "18",
        "minute": "30",
      },
      "year": "2025",
    },
    {
      "date": {
        "day": "9",
        "month": "3",
      },
      "time": {
        "hour": "16",
        "minute": "00",
      },
      "year": "2025",
    },
    {
      "date": {
        "day": "17",
        "month": "3",
      },
      "time": {
        "hour": "10",
        "minute": "00",
      },
      "year": "2025",
    },
  ],
  "timeSlotMinutes": 30,
  "title": "QuestMaker featuring サンリオキャラクターズ",
  "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
}
`;

exports[`useScheduleShare > 実際のデータでテスト > QM2のデータが正しく圧縮できること 2`] = `
[
  {
    "schedule": {
      "date": {
        "day": 2,
        "month": 3,
        "year": 2025,
      },
      "time": {
        "hour": 18,
        "minute": 30,
      },
    },
    "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
  },
  {
    "schedule": {
      "date": {
        "day": 9,
        "month": 3,
        "year": 2025,
      },
      "time": {
        "hour": 16,
        "minute": 0,
      },
    },
    "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
  },
  {
    "schedule": {
      "date": {
        "day": 17,
        "month": 3,
        "year": 2025,
      },
      "time": {
        "hour": 10,
        "minute": 0,
      },
    },
    "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
  },
]
`;

exports[`useScheduleShare > 実際のデータでテスト > QM2のデータが正しく圧縮できること 3`] = `"n1YirARqIMiNgIA8DInYCA0gIA"`;

exports[`useScheduleShare > 実際のデータでテスト > QM2のデータが正しく圧縮できること 4`] = `
[
  {
    "schedule": {
      "date": {
        "day": 2,
        "month": 3,
        "year": 2025,
      },
      "time": {
        "hour": 18,
        "minute": 30,
      },
    },
    "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
  },
  {
    "schedule": {
      "date": {
        "day": 9,
        "month": 3,
        "year": 2025,
      },
      "time": {
        "hour": 16,
        "minute": 0,
      },
    },
    "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
  },
  {
    "schedule": {
      "date": {
        "day": 17,
        "month": 3,
        "year": 2025,
      },
      "time": {
        "hour": 10,
        "minute": 0,
      },
    },
    "uid": "5a6fb727-cd94-5647-9329-b0129edfd4e5",
  },
]
`;
